[JSC] Fix references in new wasm BBQ

Constellation opened the pull request:
#### dda75455221d063750b4a33c49961155750ef774
<pre>
[JSC] Fix references in new wasm BBQ
<a href="https://bugs.webkit.org/show_bug.cgi?id=252529">https://bugs.webkit.org/show_bug.cgi?id=252529</a>
rdar://105628460

Reviewed by Mark Lam.

1. wasm references are JS values. So use uint64_t instead of pointers.
2. call_indirect should throw BadSignature error when signature is wrong.
3. addRefIsNull was using loadIfNecessary, but it does not work when value is constant. This patch changes it to use EMIT_UNARY.
   In subsequent patches, we need to scan loadIfNecessary and fix this issue since there are uses of loadIfNecessary without checking
   isConstant.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::Value::asRef const):
(JSC::Wasm::BBQJIT::Value::fromRef):
(JSC::Wasm::BBQJIT::Value::size const):
(JSC::Wasm::BBQJIT::addConstant):
(JSC::Wasm::BBQJIT::addRefIsNull):
(JSC::Wasm::BBQJIT::addRefAsNonNull):
(JSC::Wasm::BBQJIT::addRefFunc):
(JSC::Wasm::BBQJIT::addTopLevel):
(JSC::Wasm::BBQJIT::addCallIndirect):
(JSC::Wasm::BBQJIT::emitStoreConst):
(JSC::Wasm::BBQJIT::emitMoveConst):

Canonical link: <a href="https://commits.webkit.org/260516@main">https://commits.webkit.org/260516@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ced385f9912fb92279a42dbd97e41cc3cb0be8b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108497 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17594 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41347 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117603 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/117807 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19045 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8872 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100725 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114264 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14273 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97499 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42239 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96251 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29139 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97643 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10404 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30488 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/98534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8524 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11161 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7399 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/98534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16551 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50084 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106067 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12749 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26273 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3959 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5717,6 +5671,55 @@ class BBQJIT {
+            "RefIsNull", TypeKind::I32,
+            BLOCK(Value::fromI32(operand.asRef() == JSValue::encode(jsNull()))),
+            BLOCK(
+                ASSERT(JSValue::encode(jsNull()) >= 0 && JSValue::encode(jsNull()) <= INT32_MAX);
```
Can we make `JSValue::encode` and `jsNull()` constexpr and turn this into a `static_assert` instead?  Maybe can do that in a subsequent patch.

Constellation replied:
Unfortunately, it is union access, and we cannot make it constexpr.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5717,6 +5671,55 @@ class BBQJIT {
+
+        result = topValue(TypeKind::Ref);
+        Location resultLocation = allocate(result);
+        ASSERT(JSValue::encode(jsNull()) >= 0 && JSValue::encode(jsNull()) <= INT32_MAX);
```
Ditto: make `static_assert` if possible.

Constellation replied:
Ditto.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5815,19 +5818,21 @@ class BBQJIT {
+            case TypeKind::RefNull:
+            case TypeKind::Structref:
+            case TypeKind::Arrayref:
+                emitStoreConst(Value::fromI64(bitwise_cast<uint64_t>(JSValue::encode(jsNull()))), m_locals[i]);
```
We already have an `encodedJSUndefined()`.  I think the fact that we're doing this encoding and casting so often suggest that we should add an `encodedJSNull()`.

Constellation replied:
I feel `JSValue::encode(jsNull())` is more clear in terms of reading. This kind of form is used in DFG / FTL etc. everywhere, and it looks very obvious what it returns.

MenloDorian reviewed (approved):
r=m with suggestions.

Constellation committed:
[JSC] Fix references in new wasm BBQ
https://bugs.webkit.org/show_bug.cgi?id=252529
rdar://105628460

Reviewed by Mark Lam.

1. wasm references are JS values. So use uint64_t instead of pointers.
2. call_indirect should throw BadSignature error when signature is wrong.
3. addRefIsNull was using loadIfNecessary, but it does not work when value is constant. This patch changes it to use EMIT_UNARY.
   In subsequent patches, we need to scan loadIfNecessary and fix this issue since there are uses of loadIfNecessary without checking
   isConstant.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::Value::asRef const):
(JSC::Wasm::BBQJIT::Value::fromRef):
(JSC::Wasm::BBQJIT::Value::size const):
(JSC::Wasm::BBQJIT::addConstant):
(JSC::Wasm::BBQJIT::addRefIsNull):
(JSC::Wasm::BBQJIT::addRefAsNonNull):
(JSC::Wasm::BBQJIT::addRefFunc):
(JSC::Wasm::BBQJIT::addTopLevel):
(JSC::Wasm::BBQJIT::addCallIndirect):
(JSC::Wasm::BBQJIT::emitStoreConst):
(JSC::Wasm::BBQJIT::emitMoveConst):

Canonical link: https://commits.webkit.org/260516@main
